modules/pico_dns_common.c

```diff
@@ -109,66 +109,79 @@ pico_dns_namelen_comp( char *name )
  *  @param packet Packet that contains the compressed name
  *  @return Returns the decompressed name, NULL on failure.
  * ****************************************************************************/
-char *
-pico_dns_decompress_name( char *name, pico_dns_packet *packet )
-{
-    char decompressed_name[PICO_DNS_NAMEBUF_SIZE] = {
-        0
-    };
-    char *return_name = NULL;
-    uint16_t ptr = 0, nslen = 0;
-    uint16_t decompressed_index = 0;
-    char *label = NULL, *next = NULL;
-
-    /* Reading labels until reaching to pointer or NULL terminator.
-     * Only one pointer is allowed in DNS compression, the pointer is always the last according to the RFC */
-    dns_name_foreach_label_safe(label, name, next, PICO_DNS_NAMEBUF_SIZE) {
-
-        uint8_t label_size = (uint8_t)(*label+1);
-        if (decompressed_index + label_size >= PICO_DNS_NAMEBUF_SIZE) {
-            return NULL;
-        }
-        memcpy(&decompressed_name[decompressed_index], label, label_size);
-        decompressed_index = (uint16_t)(decompressed_index+label_size);
-    }
+char *pico_dns_decompress_name(char *name, pico_dns_packet *packet) {
+  uint16_t packet_len = sizeof(pico_dns_packet);
+  char decompressed_name[PICO_DNS_NAMEBUF_SIZE] = {0};
+  char *return_name = NULL;
+  uint16_t ptr = 0, nslen = 0;
+  uint16_t decompressed_index = 0;
+  char *label = NULL, *next = NULL;
+
+  if (!name || !packet) {
+    pico_err = PICO_ERR_EINVAL;
+    return NULL;
+  }
 
-    if (decompressed_index >= PICO_DNS_NAMEBUF_SIZE) {
-        return NULL;
+  /* Reading labels until reaching to pointer or NULL terminator.
+   * Only one pointer is allowed in DNS compression, the pointer is always the
+   * last according to the RFC */
+  dns_name_foreach_label_safe(label, name, next, PICO_DNS_NAMEBUF_SIZE) {
+    if (!label || (*label & 0xFF) >= PICO_DNS_NAMEBUF_SIZE) {
```

Comment:
@Diebbo Are these checks necessary?
- `name` is not a NULL pointer (with your previously added check) so `label` will not be a NULL pointer as well
- Macro `dns_name_foreach_label_safe` ensures `*label <= 63` so, to me, `(*label & 0xFF) >= PICO_DNS_NAMEBUF_SIZE)` will be always false

Let me know if I missed something or if you have an example that makes these checks relevant